File: leetcode/python/1590.make-sum-divisible-by-p.py
class Solution:
    def minSubarray(self, nums: List[int], p: int) -> int:
        rem = sum(nums) % p
        if rem == 0:
            return 0
        prefix = 0
        idxs = {0: -1}
        ans = len(nums)
        for i, n in enumerate(nums):
            prefix = (prefix + n) % p
            target = (prefix - rem + p) % p
            if target in idxs:
                ans = min(ans, i - idxs[target])
            idxs[prefix] = i
        return -1 if ans == len(nums) else ans

File: leetcode/cpp/1590.make-sum-divisible-by-p.cpp
class Solution {
public:
    int minSubarray(vector<int>& nums, int p) {
        long int sum = 0;
        for (const int n: nums) {
            sum += n;
        }
        long int rem = sum % p;
        if (rem == 0) {
            return 0;
        }
        long int prefix = 0, target = 0;
        unordered_map<long int, int> idxs;
        idxs[0] = -1;

        int ans = nums.size();

        for (int i = 0; i < nums.size(); i++) {
            prefix = (prefix + nums[i]) % p;
            target = (prefix - rem + p) % p;
            if (idxs.find(target) != idxs.cend()) {
                ans = min(ans, i - idxs[target]);
            }
            idxs[prefix] = i;
        }

        return ans == nums.size() ? -1 : ans;
    }
};

